File: 0040-combination-sum-ii/0040-combination-sum-ii.java
class Solution {
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        List<List<Integer>> res = new ArrayList<>();
        Arrays.sort(candidates);
        backtrack(res, candidates, 0, target, new ArrayList<>());
        return res;
    }

    public void backtrack(List<List<Integer>> res, int[] nums, int start, int remain, List<Integer> path){
        if(remain==0){
            res.add(new ArrayList<>(path));
            return;
        }

        for(int i=start; i<nums.length; i++){
            if(i>start && nums[i]==nums[i-1]) continue;
            if(nums[i]>remain) break;

            path.add(nums[i]);
            backtrack(res, nums, i+1, remain-nums[i], path);
            path.remove(path.size()-1);
        }
    }
}

File: 0040-combination-sum-ii/0040-combination-sum-ii.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        res = []
        candidates.sort()
        def backtrack(cur, i, total):
            if total == target:
                res.append(cur.copy())
                return
            
            for j in range(i,len(candidates)):
                if total + candidates[j] > target:
                    break
                
                if j > i and candidates[j] == candidates[j-1]:
                    continue
                
                cur.append(candidates[j])
                backtrack(cur, j+1, total + candidates[j])
                cur.pop()

        backtrack([], 0, 0)
        return res